### PR TITLE
Use versioned Perl shebangs on macOS and support replacing shebangs during relocation

### PR DESCRIPTION
**Pull request by Bo98**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

See #11227 and #11275.

Commits are split to make it easier to review.

### CODE REVIEW

File: Library/Homebrew/os/mac.rb [hunk 1]
```diff
@@ -60,6 +60,15 @@ def prerelease?
       version >= "12"
     end
 
+    sig { returns(String) }
+    def preferred_perl_version
+      if version >= :big_sur
+        "5.30"
+      else
+        "5.18"
+      end
```
Comment (Bo98) [inline Library/Homebrew/os/mac.rb:69]: We hacve the option to use 5.28 on Catalina if we want. (5.18 is the default.)

**Comment by Bo98** [comment]
Due to the nature of changing relocation, there is a window between this potentially being merged and a new brew tag being released where bottles will be created in our CI with `@@HOMEBREW_PERL@@`, but old users on the stable release channel won't be able to use such bottles.

File: Library/Homebrew/os/mac.rb (hunk 1 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/os/mac.rb:69]: @Bo98 May as well?

**Review by MikeMcQuaid — DISMISSED** [review]
Makes sense. Suggestion: when we're ready to go with this we run a single test PR through homebrew/core and then make a tag before publishing any bottles.

**Comment by MikeMcQuaid** [comment]
Great work @Bo98!

**Comment by Bo98** [comment]
Added a tweak to address #11108.

File: Library/Homebrew/keg_relocate.rb [hunk 2]
```diff
@@ -6,12 +6,43 @@ class Keg
...
   REPOSITORY_PLACEHOLDER = "@@HOMEBREW_REPOSITORY@@"
   LIBRARY_PLACEHOLDER = "@@HOMEBREW_LIBRARY@@"
+  PERL_PLACEHOLDER = "@@HOMEBREW_PERL@@"
 
-  Relocation = Struct.new(:old_prefix, :old_cellar, :old_repository, :old_library,
-                          :new_prefix, :new_cellar, :new_repository, :new_library) do
-    # Use keyword args instead of positional args for initialization.
-    def initialize(**kwargs)
-      super(*members.map { |k| kwargs[k] })
+  class Relocation
```
Comment (MikeMcQuaid) [inline Library/Homebrew/keg_relocate.rb:11]: Worth having any basic unit tests for this class?

**Review by MikeMcQuaid — APPROVED** [review]

**Comment by carlocab** [comment]
Note to self: follow up on cleanup comments from https://github.com/Homebrew/brew/pull/11331#discussion_r627237000.

**Pull request merged**